src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/matrix_nms/matrix_nms_kernel_ref.cpp

```diff
@@ -0,0 +1,184 @@
+        dispatch_data.lws = {1, 1, 1};
+    }
+
+    return dispatch_data;
+}
+}  // anonymous namespace
+
+KernelsData MatrixNmsKernelRef::GetKernelsData(const Params& params, const optional_params& options) const {
+    if (!Validate(params, options)) {
+        return {};
+    }
+
+    const size_t kernels_num{3};
+    KernelData kernel_data = KernelData::Default<matrix_nms_params>(params, kernels_num);
+    const matrix_nms_params& new_params = dynamic_cast<const matrix_nms_params&>(*kernel_data.params.get());
+
+    const size_t BOX_INFO_SIZE{16};
+
+    const int batches_num = new_params.inputs[1].Batch().v;
+    const int classes_num = new_params.inputs[1].Feature().v;
+    const int boxes_num = new_params.inputs[0].Feature().v;
+
+    int max_boxes_per_class{boxes_num};
+    if (new_params.nms_top_k >= 0)
+        max_boxes_per_class = std::min(max_boxes_per_class, new_params.nms_top_k);
+
+    auto classes_num_adj = classes_num;
+    if (new_params.background_class >= 0 && new_params.background_class < classes_num)
+        classes_num_adj = std::max(1, classes_num - 1);
+
+    auto max_boxes_per_batch = max_boxes_per_class * classes_num_adj;
+    if (new_params.keep_top_k >= 0)
+        max_boxes_per_batch = std::min(max_boxes_per_batch, new_params.keep_top_k);
+
+    const size_t box_info_num = batches_num * classes_num * max_boxes_per_class;
+
+    const size_t box_info_buffer_size = box_info_num * BOX_INFO_SIZE;
+    const size_t sel_boxes_num_buffer_size = batches_num * classes_num * sizeof(int);
+
+    kernel_data.internalBufferSizes.push_back(box_info_buffer_size);
+    kernel_data.internalBufferSizes.push_back(sel_boxes_num_buffer_size);
+    kernel_data.internalBufferDataType = Datatype::F32;
+
+    for (size_t i{}; i < kernels_num; ++i) {
+        auto entry_point = GetEntryPoint(kernelName, new_params.layerID, params, options, i);
+        auto jit_constants = GetJitConstants(new_params);
+        jit_constants.AddConstant(MakeJitConstant("MATRIX_NMS_STAGE_" + std::to_string(i), "true"));
+
+        jit_constants.AddConstant(MakeJitConstant("MAX_BOXES_PER_CLASS", max_boxes_per_class));
+        jit_constants.AddConstant(MakeJitConstant("MAX_BOXES_PER_BATCH", max_boxes_per_batch));
+        auto jit = CreateJit(kernelName, jit_constants, entry_point);
+
+        DispatchData dispatch_data = SetDefault(new_params, i);
+        auto& kernel = kernel_data.kernels[i];
+        KernelBase::CheckDispatchData(kernelName, dispatch_data, params.engineInfo.maxWorkGroupSize);
+        kernel.params.workGroups.global = dispatch_data.gws;
+        kernel.params.workGroups.local = dispatch_data.lws;
+        kernel.code.kernelString = GetKernelString(kernelName, jit, entry_point, params.engineInfo);
+
+        SetKernelArguments(new_params, kernel, i);
+    }
+
+    return {kernel_data};
+}
+
+float MatrixNmsKernelRef::GetKernelsPriority(const Params& params, const optional_params& options) const {
+    return FORCE_PRIORITY_1;
```

Comment:
I thought reference kernels should have the lowest priority, not the biggest ones, no?

Reply:
This is a good question! @tgubanova-lohika could you help, what priority should we have here?

Reply:
usually ref kernels should be with lowest priority: FORCE_PRIORITY_9 or DONT_USE_IF_HAVE_SOMETHING_ELSE, mostly 9 is used as I see

Reply:
done